File: webflux-dialogue/src/main/java/com/study/webflux/rag/infrastructure/monitoring/adapter/MongoPerformanceMetricsRepository.java
package com.study.webflux.rag.infrastructure.monitoring.adapter;

import java.time.Instant;

import lombok.RequiredArgsConstructor;

import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Repository;

import com.study.webflux.rag.domain.monitoring.model.PerformanceMetrics;
import com.study.webflux.rag.domain.monitoring.port.PerformanceMetricsRepository;
import com.study.webflux.rag.infrastructure.monitoring.document.PerformanceMetricsDocument;
import com.study.webflux.rag.infrastructure.monitoring.repository.SpringDataPerformanceMetricsRepository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Repository
@RequiredArgsConstructor
public class MongoPerformanceMetricsRepository implements PerformanceMetricsRepository {

	private final SpringDataPerformanceMetricsRepository repository;

	@Override
	public Mono<PerformanceMetrics> save(PerformanceMetrics metrics) {
		return Mono.just(metrics)
			.map(PerformanceMetricsDocument::fromDomain)
			.flatMap(repository::save)
			.map(PerformanceMetricsDocument::toDomain);
	}

	@Override
	public Mono<PerformanceMetrics> findById(String pipelineId) {
		return repository.findById(pipelineId)
			.map(PerformanceMetricsDocument::toDomain);
	}

	@Override
	public Flux<PerformanceMetrics> findByTimeRange(Instant startTime, Instant endTime) {
		return repository.findByStartedAtBetweenOrderByStartedAtDesc(startTime, endTime)
			.map(PerformanceMetricsDocument::toDomain);
	}

	@Override
	public Flux<PerformanceMetrics> findByStatus(String status, int limit) {
		int pageSize = Math.max(1, limit);
		return repository.findByStatusOrderByStartedAtDesc(status, PageRequest.of(0, pageSize))
			.map(PerformanceMetricsDocument::toDomain);
	}

	@Override
	public Flux<PerformanceMetrics> findSlowPipelines(long thresholdMillis, int limit) {
		int pageSize = Math.max(1, limit);
		return repository.findSlowPipelines(thresholdMillis, PageRequest.of(0, pageSize))
			.map(PerformanceMetricsDocument::toDomain);
	}

	@Override
	public Flux<PerformanceMetrics> findRecent(int limit) {
		int pageSize = Math.max(1, limit);
		return repository.findAllByOrderByStartedAtDesc(PageRequest.of(0, pageSize))
			.map(PerformanceMetricsDocument::toDomain);
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/infrastructure/monitoring/adapter/MongoPerformanceMetricsRepository.kt
package com.miyou.app.infrastructure.monitoring.adapter

import com.miyou.app.domain.monitoring.model.PerformanceMetrics
import com.miyou.app.domain.monitoring.port.PerformanceMetricsRepository
import com.miyou.app.infrastructure.monitoring.document.PerformanceMetricsDocument
import com.miyou.app.infrastructure.monitoring.repository.SpringDataPerformanceMetricsRepository
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Repository
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.Instant

@Repository
class MongoPerformanceMetricsRepository(
    private val repository: SpringDataPerformanceMetricsRepository,
) : PerformanceMetricsRepository {
    override fun save(metrics: PerformanceMetrics): Mono<PerformanceMetrics> =
        Mono
            .just(metrics)
            .map(PerformanceMetricsDocument::fromDomain)
            .flatMap(repository::save)
            .map(PerformanceMetricsDocument::toDomain)

    override fun findById(pipelineId: String): Mono<PerformanceMetrics> =
        repository
            .findById(pipelineId)
            .map(PerformanceMetricsDocument::toDomain)

    override fun findByTimeRange(
        startTime: Instant,
        endTime: Instant,
    ): Flux<PerformanceMetrics> =
        repository
            .findByStartedAtBetweenOrderByStartedAtDesc(startTime, endTime)
            .map(PerformanceMetricsDocument::toDomain)

    override fun findByStatus(
        status: String,
        limit: Int,
    ): Flux<PerformanceMetrics> {
        val pageSize = limit.coerceAtLeast(1)
        return repository
            .findByStatusOrderByStartedAtDesc(status, PageRequest.of(0, pageSize))
            .map(PerformanceMetricsDocument::toDomain)
    }

    override fun findSlowPipelines(
        thresholdMillis: Long,
        limit: Int,
    ): Flux<PerformanceMetrics> {
        val pageSize = limit.coerceAtLeast(1)
        return repository
            .findByTotalDurationMillisGreaterThanEqual(
                thresholdMillis,
                PageRequest.of(0, pageSize, Sort.by(Sort.Direction.DESC, "totalDurationMillis")),
            ).map(PerformanceMetricsDocument::toDomain)
    }

    override fun findRecent(limit: Int): Flux<PerformanceMetrics> {
        val pageSize = limit.coerceAtLeast(1)
        return repository
            .findAllByOrderByStartedAtDesc(PageRequest.of(0, pageSize))
            .map(PerformanceMetricsDocument::toDomain)
    }
}
